src/NHibernate/Transform/DistinctRootEntityResultTransformer.cs

```diff
@@ -77,13 +72,10 @@ public bool IsTransformedValueATupleElement(String[] aliases, int tupleLength)
 
 		public override bool Equals(object obj)
 		{
-			if (obj == null || obj.GetHashCode() != Hasher.GetHashCode())
-			{
-				return false;
-			}
-			// NH-3957: do not rely on hashcode alone.
-			// Must be the exact same type
-			return obj.GetType() == typeof(DistinctRootEntityResultTransformer);
+			if (ReferenceEquals(obj, this))
+				return true;
+
+			return obj is DistinctRootEntityResultTransformer;
```

Comment:
This is not equivalent because the class is not sealed.

Reply:
this equality issue affects all transformers. I'm going to fix it in separate pr